Sources/SwiftCrossUI/Views/NavigationPath.swift

```diff
@@ -16,17 +27,93 @@ public struct NavigationPath {
     public init() {}
 
     /// Appends a new value to the end of this path.
-    public mutating func append<H: Hashable>(_ component: H) {
-        path.append(component)
+
+    public mutating func append<C: Codable>(_ component: C) {
+        entries.append(Entry(
+            type: typeName(for: C.self),
+            data: component
+        ))
     }
 
     /// Removes values from the end of this path.
     public mutating func removeLast(_ k: Int = 1) {
-        path.removeLast(k)
+        entries.removeLast(k)
     }
 
     /// Removes all values from this path.
     public mutating func removeAll() {
-        path.removeAll()
+        entries.removeAll()
+    }
+
+    internal func afterDecodingEntries<T: Codable>(ofType type: T.Type) -> NavigationPath? {
+        guard hasPendingDecodableEntries else {
+            return nil
+        }
+
+        let name = typeName(for: type)
+        var didDecodeAnEntry = false
+        var didFindDecodableEntry = false
+        var updated = self
+        updated.entries = entries.map {
+            if let decodable = $0 as? LazyDecodableEntry {
+                didFindDecodableEntry = true
+                if decodable.type == name, let data = try? JSONDecoder().decode(type, from: decodable.data) {
+                    didDecodeAnEntry = true
+                    return Entry(type: name, data: data)
+                }
+            }
+
+            return $0
+        }
+
+        if didDecodeAnEntry || !didFindDecodableEntry {
+            updated.hasPendingDecodableEntries = didFindDecodableEntry
+            return updated
+        }
+
+        return nil
+    }
+
+    private func typeName<T>(for: T.Type) -> String {
+        // If this also does not work we can do String(describing:) but that wont
+        // include parent of nested class and module
```

Comment:
I haven't tested _typeName yet, but if `_typeName` isn't available `String(reflecting:)` is what we want I think, it gives you `NavigationExample.Science.Subject` for example.

Reply:
_typeName should be available